lib/puma/launcher.rb

```diff
@@ -286,6 +294,18 @@ def prune_bundler
       end
     end
 
+    def spec_for_gem(gem_name)
+      Bundler.rubygems.loaded_specs(gem_name)
+    end
+
+    def require_paths_for_gem(gem_spec)
```

Comment:
So, we're actually requiring Rubygems 2.2+ *to use prune_bundler* at all now, is that correct?

I'm actually OK with that as well (again, rubygems 2.2+ is older than ruby 2.2) but we need to gate the code on that as well.

Reply:
Oh I see, you've done a fallback instead. 

I'm wondering if we should even do that. I think I sort of proved in the comments of this PR that the original implementation was bugged.

Reply:
I was just planning on straight-up raising on attempts to use the feature without rubygems 2.2+ (https://github.com/puma/puma/pull/1105/commits/85c7f21cff603b1e65e2e2dadb620ae47b72001e#diff-d9e00ab028817951ecfdbbe7e661a141R471) instead of silently failing. The fallback codepath here is because this method is now used to find the puma lib dir. Previously it was just the fallback line, run inline. Now it's the preferred codepath for clients with rubygems 2.2+ but we still support the old codepath.

Reply:
The problem is that the old codepath is clearly bugged.

Try this script and you'll see what I mean: 

```ruby
require 'bundler'
require 'puma'
gem_spec = Bundler.rubygems.loaded_specs("puma")
puts gem_spec.full_require_paths
puts gem_spec.require_paths.map { |x| File.join(gem_spec.full_gem_path, x) }
```

The latter line, for me, outputs nonsense like: 

```
"/Users/nateberkopec/.gem/ruby/2.6.3/gems/puma-4.0.1/Users/nateberkopec/.gem/ruby/2.6.3/extensions/x86_64-darwin-18/2.6.0-static/puma-4.0.1"
```

...which does not exist.

There's not really a point in maintaining a 2nd codepath that doesn't actually work, so we need to either remove the path or fix it.

Reply:
Puma's requiring Ruby >= 2.2.0, which shipped with [RubyGems 2.4.5](https://github.com/ruby/ruby/blob/v2_2_0/lib/rubygems.rb#L12)

Reply:
I'm unfamiliar with what "shipped with" means _exactly_ in this context. Is it possible that _anyone_ is using Ruby 2.2 or greater with Rubygems 2.3 or less?

Reply:
Sorry.  It's possible, but Ruby 2.2.0 included RubyGems 2.4.5 as a std-lib.

'Shipped with' <=> 'included'

Reply:
Right. Well I'd rather provide a feature that works over one that doesn't, so our choices are to: 

a) copy the entire implementation of `full_require_paths` to this file and then remove the rubygems gate 
b) remove the branch, and gate the entire feature of prune_bundler on Rubygems 2.2+

I am fine with either option.